packages/router/src/grpc/view-protocol-server/nullifier-status.ts

```diff
@@ -1,91 +1,58 @@
-  return noteSpent || swapSpent;
-}
+export const nullifierStatus: Impl['nullifierStatus'] = async (req, ctx) => {
+  const hasWallet = ctx.values.get(hasWalletCtx);
+  await hasWallet(req.walletId);
 
-export const handleNullifierStatusReq = async (
-  req: NullifierStatusRequest,
-  services: ServicesInterface,
-): Promise<NullifierStatusResponse> => {
-  await assertWalletIdMatches(req.walletId);
-  if (!req.nullifier) throw new Error('No nullifier passed');
+  const { nullifier } = req;
+  if (!nullifier) throw new Error('No nullifier passed');
 
+  const services = ctx.values.get(servicesCtx);
   const { indexedDb } = await services.getWalletServices();
-  const spent = await noteOrSwapSpent(indexedDb, req.nullifier);
-
-  if (spent) {
-    return new NullifierStatusResponse({ spent: true });
-  } else if (!req.awaitDetection) {
-    return new NullifierStatusResponse({ spent: false });
-  }
-
-  // Wait until our DB encounters a new note or swap. If it corresponds to the nullifier, break loop.
-  const mergedSubscription = mergeAsyncGenerators(
-    indexedDb.subscribe('SPENDABLE_NOTES'),
-    indexedDb.subscribe('SWAPS'),
-  );
 
-  for await (const update of mergedSubscription) {
-    if (update.table === 'SPENDABLE_NOTES') {
-      const note = SpendableNoteRecord.fromJson(update.value);
-      if (note.nullifier?.equals(req.nullifier) && note.heightSpent !== 0n) break;
-    } else {
-      const swap = SwapRecord.fromJson(update.value);
-      if (swap.nullifier?.equals(req.nullifier) && swap.heightClaimed !== 0n) break;
-    }
+  // grab subscription before local check. this avoids race condition
+  const swapStream = indexedDb.subscribe('SWAPS');
+  const noteStream = indexedDb.subscribe('SPENDABLE_NOTES');
+
+  // If present, a swap or note should never have an undefined height, and a
+  // zero-height spend should never appear. So if one of these is truthy, the
+  // nullifier is spent.
+  const [swap, note] = await Promise.all([
+    indexedDb.getSwapByNullifier(nullifier),
+    indexedDb.getNoteByNullifier(nullifier),
+  ]);
+  const spent = Boolean(swap?.heightClaimed) || Boolean(note?.heightSpent);
+
+  if (!spent && req.awaitDetection) {
+    // use of the nullifier was not present in local data, so watch new data.
+    // we might check a few very recent items twice, if they were included in
+    // local check, but we won't miss anything.
+    const eventuallySpent = Promise.race([
+      watchStream(swapStream, ({ value: swapJson }) => {
+        const swap = SwapRecord.fromJson(swapJson);
+        return Boolean(swap.heightClaimed) && nullifier.equals(swap.nullifier);
+      }),
+      watchStream(noteStream, ({ value: noteJson }) => {
+        const note = SpendableNoteRecord.fromJson(noteJson);
+        return Boolean(note.heightSpent) && nullifier.equals(note.nullifier);
+      }),
+    ]);
+
+    return eventuallySpent.then(() => ({ spent: true }));
```

Comment:
@grod220 used Promise.race to independently iterate, rather than attempting to merge asynciterators